cpp/src/strings/split/split.cu

```diff
@@ -385,6 +386,428 @@ struct whitespace_rsplit_tokenizer_fn
+    for (size_type i = 0; i < d_str.length(); ++i) {
+      auto const cur_pos = forward ? i : d_str.length() - 1 - i;
+      auto const ch = d_str[cur_pos];
+      if (spaces != (ch <= ' ')) {
+        if (spaces) {  // from whitespace(s) to a new token
+          to_token_pos = cur_pos;
+        }
+        else {  // from a token to whiltespace(s)
+          if (token_count < max_tokens - 1) {
+            token_count++;
+            token_size_sum +=
+              compute_token_char_bytes<false>(d_str, cur_pos, to_token_pos);
+          }
+          else {
+            reached_max_tokens = true;
+            break;
+          }
+        }
+        spaces = !spaces;
+      }
+    }
+    if (reached_max_tokens || !spaces) {
+      token_count++;
+      token_size_sum +=
+        compute_token_char_bytes<true>(d_str, -1, to_token_pos);
+    }
+
+    if (token_count == 0) {  // note that pandas.Series.str.split("", pat=" ")
+                             // returns one token (i.e. "") while
+                             // pandas.Series.str.split("") returns 0 token.
+      return thrust::make_tuple<size_type, size_type, size_type>(0, 0, 0);
+    }
+
+    auto const memory_size = compute_memory_size(token_count, token_size_sum);
+
+    return thrust::make_tuple<size_type, size_type, size_type>(
+      token_count, token_size_sum, memory_size);
+  }
+};
+
+/**
+ * @brief Copy the tokens from the `idx'th` string element of `d_strings` to
+ * the contiguous memory buffer.
+ */
+template <bool forward>
+struct whitespace_token_copier_fn {
+  column_device_view const d_strings;  // strings to split
+  bool const has_validity = false;
+
+  template <bool last>
+  __device__ thrust::pair<size_type, size_type>
+  compute_src_byte_offset_and_token_char_bytes(
+      string_view const& d_str,
+      size_type cur_pos, size_type to_token_pos,
+      size_type remaining_bytes) const {
+    if (last) {
+      auto const token_char_bytes = remaining_bytes;
+      auto const src_byte_offset =
+        forward ? d_str.byte_offset(to_token_pos) :
+                  d_str.byte_offset(to_token_pos + 1) - token_char_bytes;
+      return thrust::make_pair<size_type, size_type>(
+        src_byte_offset, token_char_bytes);
+    }
+    else {
+      auto const src_byte_offset =
+        forward ? d_str.byte_offset(to_token_pos) :
+          d_str.byte_offset(cur_pos + 1);
+      auto const token_char_bytes =
+        forward ?
+          d_str.byte_offset(cur_pos) - src_byte_offset :
+          d_str.byte_offset(to_token_pos + 1) - src_byte_offset;
+      return thrust::make_pair<size_type, size_type>(
+        src_byte_offset, token_char_bytes);
+    }
+  }
+
+  __device__ void
+  operator()(
+      thrust::tuple<size_type, size_type, void*, size_type>
+        idx_token_count_memory_ptr_token_size_sum) const {
+    auto const idx = idx_token_count_memory_ptr_token_size_sum.get<0>();
+    auto const token_count = idx_token_count_memory_ptr_token_size_sum.get<1>();
```

Comment:
We have to be able to better than this. Having a 4 element tuple and naming it `idx_token_count_memory_ptr_token_size_sum` is confusing and less than ideal naming. You have to parse that into `idx` `token count` `memory_ptr` and `token size sum`. Also, you have to remember the order and use `get<#>` each time which is also not very readable. Does this have to be a `thrust::tuple`? Can it be a `struct`? aka
```cpp
struct copy_info { // not sure if this is the best name
    size_type idx;
    size_type token_count;
    size_type token_size_sum;
    void* memory_ptr;
};
```